backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportHeaderGenerator.kt

```diff
@@ -973,3 +1018,50 @@ private fun ObjCExportNamer.ClassOrProtocolName.toNameAttributes(): List<String>
 
 private fun swiftNameAttribute(swiftName: String) = "swift_name(\"$swiftName\")"
 private fun objcRuntimeNameAttribute(name: String) = "objc_runtime_name(\"$name\")"
+
+interface TypeParameterProvider{
+    fun typeAvailable(kotlinType: KotlinType):Boolean
+}
+
+internal class ClassParameterProvider(container:DeclarationDescriptor): TypeParameterProvider {
+    constructor(method: FunctionDescriptor) : this(method.containingDeclaration)
+    private val nameSet:Set<String>
+    init {
+        nameSet = mutableSetOf()
+        if(container is ClassDescriptor && !container.isInterface) {
+            container.declaredTypeParameters.forEach {
+                nameSet.add(it.name.asString())
+            }
+        }
+    }
+
+    override fun typeAvailable(kotlinType: KotlinType): Boolean =
+            nameSet.contains(kotlinType.typeParameterName())
+}
+
+internal object NoneTypeParameterProvider: TypeParameterProvider{
+    override fun typeAvailable(kotlinType: KotlinType): Boolean = false
+}
+
+
+private fun Variance.objcDeclaration():String = when(this){
+    Variance.OUT_VARIANCE -> "__covariant "
+    Variance.IN_VARIANCE -> "__contravariant "
+    else -> ""
+}
+
+//To anybody reviewing this code, I don't know my way
+//around the Kotlin parsed type system too well. What
+//I'm looking for here is generic type name minus the '?'
+//I'm sure there's a much simpler way to get that
```

Comment:
Type parameters are represented with `TypeParameterDescriptor` in Kotlin compiler frontend.
You can get one using `TypeUtils.getTypeParameterDescriptorOrNull(this)`.

Reply:
Updated. Thanks.